crates/wast/src/resolve/component_names.rs

```diff
@@ -0,0 +1,817 @@
+) -> Result<(), Error> {
+    match def {
+        ComponentTypeDef::DefType(d) => resolve_deftype(d, resolve_stack),
+        ComponentTypeDef::InterType(i) => resolve_intertype(i, resolve_stack),
+    }
+}
+
+fn resolve_nested_component_type<'a, 'b>(
+    c: &'b mut ComponentType<'a>,
+    resolve_stack: &'b mut Vec<ComponentResolver<'a>>,
+) -> Result<(), Error> {
+    resolve_stack.push(ComponentResolver::new(c.id));
+
+    // Iterate through the fields of the component type. We use an index
+    // instead of an iterator because we'll be inserting aliases
+    // as we go.
+    let mut i = 0;
+    while i < c.fields.len() {
+        // Resolve names within the field.
+        match &mut c.fields[i] {
+            ComponentTypeField::Alias(alias) => {
+                resolve_alias(alias, resolve_stack)?;
+            }
+            ComponentTypeField::Type(ty) => {
+                resolve_type_def(&mut ty.def, resolve_stack)?;
+            }
+            ComponentTypeField::Import(import) => {
+                resolve_deftype_use(&mut import.type_, resolve_stack)?;
+            }
+            ComponentTypeField::Export(export) => {
+                resolve_deftype_use(&mut export.item, resolve_stack)?;
+            }
+        }
+
+        // Name resolution may have emitted some aliases. Insert them before
+        // the current definition.
+        let resolver = resolve_stack.last_mut().unwrap();
+        let aliases_to_insert = replace(&mut resolver.aliases_to_insert, Vec::new());
+        for alias in aliases_to_insert {
+            c.fields.insert(i, ComponentTypeField::Alias(alias));
+            i += 1;
+        }
+
+        // Definitions can't refer to themselves or to definitions that appear
+        // later in the format. Now that we're done resolving this field,
+        // assign it an index for later defintions to refer to.
+        match &mut c.fields[i] {
+            ComponentTypeField::Alias(alias) => {
+                register_alias(alias, resolver)?;
+            }
+            ComponentTypeField::Type(ty) => {
+                resolver.types.register(ty.id, "type")?;
+            }
+            ComponentTypeField::Import(import) => match &import.type_ {
+                ComponentTypeUse::Inline(inline) => match inline {
+                    DefType::Func(f) => {
+                        resolver.funcs.register(f.id, "component func")?;
+                    }
+                    DefType::Module(m) => {
+                        resolver.modules.register(m.id, "module")?;
+                    }
+                    DefType::Component(c) => {
+                        resolver.components.register(c.id, "component")?;
+                    }
+                    DefType::Instance(i) => {
+                        resolver.instances.register(i.id, "instance")?;
+                    }
+                    DefType::Value(v) => {
+                        resolver.values.register(v.id, "value")?;
+                    }
```

Comment:
Is it necessary to maintain these index spaces for component types? I thought that the only thing that mattered at this layer was names-of-types?